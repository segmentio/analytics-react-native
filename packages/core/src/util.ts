import type { EventPlugin } from './plugin';
import type { Timeline } from './timeline';

export const chunk = (array: any[], size: number) => {
  if (!array.length || !size) {
    return [];
  }

  let index = 0;
  let resIndex = 0;
  const result = new Array(Math.ceil(array.length / size));

  while (index < array.length) {
    result[resIndex++] = array.slice(index, (index += size));
  }
  return result;
};

export const getAllPlugins = (timeline: Timeline) => {
  const allPlugins = Object.values(timeline.plugins);
  if (allPlugins.length) {
    return allPlugins.reduce((prev = [], curr = []) => prev.concat(curr));
  }
  return [];
};

export const getPluginsWithFlush = (timeline: Timeline) => {
  if (!timeline) {
    return [];
  }

  const allPlugins = getAllPlugins(timeline);

  // checking for the existence of .flush()
  const eventPlugins = allPlugins?.filter(
    (f) => (f as EventPlugin).flush
  ) as EventPlugin[];

  return eventPlugins;
};

export const getPluginsWithReset = (timeline: Timeline) => {
  if (!timeline) {
    return [];
  }

  const allPlugins = getAllPlugins(timeline);

  // checking for the existence of .reset()
  const eventPlugins = allPlugins?.filter(
    (f) => (f as EventPlugin).reset
  ) as EventPlugin[];

  return eventPlugins;
};
