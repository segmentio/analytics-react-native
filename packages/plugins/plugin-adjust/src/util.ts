import type { SegmentAdjustSettings } from '../../../core/src/types';

export const mappedCustomEventToken = (
  eventName: string,
  settings: SegmentAdjustSettings
) => {
  let result = null;
  const tokens = settings?.customEvents;
  if (tokens) {
    result = tokens[eventName];
  }
  return result;
};

export const extract = <T>(
  key: string,
  properties: { [key: string]: unknown },
  defaultValue: T | null = null
) => {
  let result = defaultValue;
  Object.entries(properties).forEach(([propKey, propValue]) => {
    // not sure if this comparison is actually necessary,
    // but existed in the old destination so ...
    if (key.toLowerCase() === propKey.toLowerCase()) {
      result = propValue as T;
    }
  });
  return result;
};
