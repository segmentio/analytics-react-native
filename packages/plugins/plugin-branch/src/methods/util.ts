import { isObject } from '@segment/analytics-react-native';
import branch, { BranchEvent } from 'react-native-branch';
import { mapEventProps, Product, transformMap } from './parameterMapping';

function toJSONString(value: unknown): string {
  if (typeof value === 'string') return value;
  return JSON.stringify(value);
}

export async function createBranchEventWithProps(
  eventName: string,
  eventProps: { [key: string]: unknown },
  isStandardBranchEvent: boolean
): Promise<BranchEvent> {
  const products =
    eventProps.product_id !== undefined
      ? transformMap.products([eventProps])
      : eventProps.products;
  const branchUniversalObjects = [] as unknown[];

  // for each product item, create a Branch Universal Object
  if (Array.isArray(products)) {
    for (const item of products as Product[]) {
      const product_id = item.contentMetadata.customMetadata.product_id;
      if (product_id !== undefined) {
        if (isStandardBranchEvent) {
          const buo = await branch.createBranchUniversalObject(
            product_id,
            item
          );
          branchUniversalObjects.push(buo);
        } else {
          eventProps.customData = {
            ...(isObject(eventProps.customData) ? eventProps.customData : {}),
            [product_id]: toJSONString(item),
          };
        }
      }
    }
  } else {
    // separate custom non-Branch entries before forwarding
    const customData = {} as { [key: string]: string };
    const branchData = {} as { [key: string]: unknown };
    for (const key in eventProps) {
      if (key in mapEventProps) branchData[key] = eventProps[key];
      customData[key] = toJSONString(eventProps[key]);
    }
    eventProps = {
      ...branchData,
      customData,
    };
  }
  return new BranchEvent(
    eventName,
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    //@ts-ignore: Branch does not export the proper types for BranchUniversalObject so we can only use unknown
    isStandardBranchEvent ? branchUniversalObjects : undefined,
    eventProps
  );
}
