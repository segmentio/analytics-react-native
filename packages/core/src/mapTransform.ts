import { isObject } from './util';

/**
 * Generates a mapTransform function from a keymap and a transform map.
 * The resulting function can be called with any object/dictionary and return
 * a new object with key replacements and value transformations applied
 *
 * @param keyMap A dictionary of key mappings in format: ['OLD':'NEW']
 * @param transformMap A map of transformations by key (new key if the key was replaced)
 * @returns A function that maps keys from an object and transforms its values
 */
export const generateMapTransform = (
  keyMap: { [key: string]: string },
  transformMap: { [key: string]: (oldValue: unknown) => unknown }
) => {
  const mapTransform = (
    json: Record<string, unknown>
  ): Record<string, unknown> => {
    // Clone at top level
    const result: Record<string, unknown> = {};

    for (const key in json) {
      const newKey = keyMap[key] ?? key;

      let value = json[key];

      if (Array.isArray(value)) {
        value = value.map((nestedValue) => {
          if (isObject(nestedValue)) {
            return mapTransform(nestedValue);
          }
          return nestedValue as unknown;
        });
      } else if (isObject(value)) {
        value = mapTransform(value);
      }

      if (newKey in transformMap) {
        value = transformMap[newKey](value);
      }

      result[newKey] = value;
    }

    return result;
  };
  return mapTransform;
};
