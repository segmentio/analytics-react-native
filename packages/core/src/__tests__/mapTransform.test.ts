import { generateMapTransform } from '../mapTransform';
import { isString } from '../util';

describe('mapTransform', () => {
  const keyMap: { [key: string]: string } = {
    oldId: 'newId',
    veryOldKey: 'muchNewerKey',
    oldArray: 'newArray',
  };

  const transformMap: { [key: string]: (value: unknown) => unknown } = {
    muchNewerKey: (value: unknown) => {
      if (isString(value)) {
        return value.slice(0, 3);
      }
      return value;
    },
    newArray: (value: unknown) => {
      if (Array.isArray(value)) {
        return value.slice(0, 1) as unknown[];
      }
      return value;
    },
  };

  const testMapTransform = generateMapTransform(keyMap, transformMap);

  it('maps keys to new values', () => {
    const oldObject = {
      oldId: 1,
      veryOldKey: '123456',
      nestedObject: {
        veryOldKey: '123456',
        nestedObject: {
          veryOldKey: '123456',
        },
      },
      oldArray: [
        {
          veryOldKey: '123456',
          doNotTransform: 'nope',
        },
        {
          veryOldKey: '123456',
          doNotTransform: 'nope',
        },
        {
          veryOldKey: '123456',
          doNotTransform: 'nope',
        },
      ],
    };

    const clone = JSON.parse(JSON.stringify(oldObject)) as object;

    const newObject = testMapTransform(oldObject);

    // Original object shouldn't have mutated
    expect(oldObject).toMatchObject(clone);
    // New object should have new keys and values transformed
    expect(newObject).toMatchSnapshot();
  });
});
