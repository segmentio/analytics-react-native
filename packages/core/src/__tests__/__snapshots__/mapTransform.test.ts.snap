// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapTransform maps keys to new values 1`] = `
Object {
  "muchNewerKey": "123",
  "nestedObject": Object {
    "muchNewerKey": "123",
    "nestedObject": Object {
      "muchNewerKey": "123",
    },
  },
  "newArray": Array [
    Object {
      "doNotTransform": "nope",
      "muchNewerKey": "123",
    },
  ],
  "newId": 1,
}
`;
